### PR TITLE
Eedeleon/release skinny client

### PR DESCRIPTION
**Pull request by eddiestudies**
## What changes are proposed in this pull request?

This PR releases the mlflow-skinny library. The MLflow Skinny library has the same code of the MLflow library with a reduced dependency set that omits server, store, and data science specific libraries such as Flask, sqlalchemy, pandas, and numpy.

## How is this patch tested?

A new build is added for mlflow-skinny

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Release `mlflow-skinny` an MLflow package with fewer baked in dependencies to allow for a reduced footprint training and deployment option for mlflow users. `mlflow` functionality can be added for store and model scenarios by adding the corresponding dependencies along with `mlflow-skinny`

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: README_SKINNY_PREFIX.rst
Comment (dbczumar) [inline README_SKINNY_PREFIX.rst:5]:
```suggestion
MLflow Skinny is a lightweight MLflow package without metadata / artifact storage, server, UI, or data science dependencies.
```

File: README_SKINNY_PREFIX.rst
Comment (dbczumar) [inline README_SKINNY_PREFIX.rst:13]:
```suggestion
Additional dependencies can be installed to leverage the full feature set of MLflow. For example:
* To use the `mlflow.sklearn` component of MLflow Models, install `scikit-learn`, `numpy`, and `pandas`.
* To use SQL-based metadata storage, install `sqlalchemy`, `alembic`, and `sqlparse`.
```

File: README_SKINNY_PREFIX.rst
Comment (dbczumar) [inline README_SKINNY_PREFIX.rst:1]: Can we rename this file to `README_SKINNY`?

File: tests/resources/mlflow-test-plugin/setup.py
Comment (dbczumar) [inline tests/resources/mlflow-test-plugin/setup.py:13]: Is this strictly necessary? We publish this test plugin example in our docs on `mlflow.org`. The less complexity / confusion we can add, the better.

**Review by dbczumar — COMMENTED** [review]
Looks pretty good to me! Thoughts @harupy ?

**Comment by harupy** [comment]
@dbczumar Looks pretty good to me too! I'm building distribution files on this PR's branch to check how large they are.

Screen Shot 2021-02-03 at 16 36 08

I think the wheel file should be much smaller.

**Comment by harupy** [comment]
Wrote a tiny script to make sure `setup.py` works as expected:

```sh
#!/usr/bin/env bash

set -x

export MLFLOW_SKINNY="true"
echo $MLFLOW_SKINNY

python setup.py --name
python setup.py --version
python setup.py --long-description | head -n 3
```

This outputs:

```
+ export MLFLOW_SKINNY=true
+ MLFLOW_SKINNY=true
+ echo true
true
+ python setup.py --name
mlflow-skinny
+ python setup.py --version
0.1.0
+ python setup.py --long-description
+ head -n 3
=============================================
MLflow Skinny: A Light Weight Machine Learning Lifecycle Platform Client
=============================================
```

File: README_SKINNY_PREFIX.rst
Comment (dbczumar) [inline README_SKINNY_PREFIX.rst:2]:
```suggestion
MLflow Skinny: A Lightweight Machine Learning Lifecycle Platform Client
```

File: tests/resources/mlflow-test-plugin/setup.py
Comment (eedeleon) [reply tests/resources/mlflow-test-plugin/setup.py:13]: With Harupy's new change it won't be

File: README_SKINNY_PREFIX.rst
Comment (eedeleon) [reply README_SKINNY_PREFIX.rst:1]: of course

**Comment by eedeleon** [comment]
Depends on https://github.com/mlflow/mlflow/pull/4054 after removing the plugin changes.

Otherwise mlflow proper is installed and skinny dependency omission tests fail

File: setup.py
Comment (dbczumar) [inline setup.py:79]: @eedeleon Do we need this conditional? Ideally, we'll use version 1.14.0 for both the core and skinny client the next time we have a release.

File: README_SKINNY.rst
Comment (dbczumar) [inline README_SKINNY.rst:9]: For some reason, `twine` complains about the RST formatting when uploading to test PyPI. I think that the spacing before / after list items may be the culprit:

```suggestion
MLflow Skinny is a lightweight MLflow package without SQL storage, server, UI, or data science dependencies.

MLflow Skinny supports:

* Tracking operations (logging / loading / searching params, metrics, tags + logging / loading artifacts
* Model registration, search, artifact loading, and transitions
* Execution of GitHub projects within notebook & against a remote target.
```

File: README_SKINNY.rst
Comment (dbczumar) [inline README_SKINNY.rst:14]: For some reason, `twine` complains about the RST formatting when uploading to test PyPI. I think that the spacing of list items may be the culprit:

```suggestion
Additional dependencies can be installed to leverage the full feature set of MLflow. For example:

* To use the `mlflow.sklearn` component of MLflow Models, install `scikit-learn`, `numpy`, and `pandas`.
* To use SQL-based metadata storage, install `sqlalchemy`, `alembic`, and `sqlparse`.
```

Comment (harupy) [reply README_SKINNY.rst:14]:
```
python -m readme_renderer README_SKINNY_PREFIX.rst -o README_SKINNY_PREFIX.html
<string>:1: (WARNING/2) Title overline too short.

=============================================
MLflow Skinny: A Light Weight Machine Learning Lifecycle Platform Client
=============================================
```

Comment (harupy) [reply README_SKINNY.rst:14]: Filed a PR to make it easier to detect issues like this one:

https://github.com/mlflow/mlflow/pull/4059/

Comment (eedeleon) [reply README_SKINNY.rst:14]: Thanks. that tool will be super helpful

**Comment by harupy** [comment]
@eedeleon @dbczumar I ran our internal Jenkins job and successfully uploaded `mlflow-skinny` on TestPyPI:

https://test.pypi.org/project/mlflow-skinny/0.1.0/

File: setup.py
Comment (eedeleon) [reply setup.py:79]: I can remove it

**Review by dbczumar — APPROVED** [review]
LGTM once tests pass! Thanks @eedeleon . @harupy let me know if you have any additional blocking feedback here before merge.

File: README_SKINNY.rst
Comment (harupy) [inline README_SKINNY.rst:8]:
```suggestion
* Tracking operations (logging / loading / searching params, metrics, tags + logging / loading artifacts)
```

nit: Is `)` missing?

File: README_SKINNY.rst
Comment (harupy) [inline README_SKINNY.rst:15]: https://github.com/mlflow/mlflow/pull/4049/checks?check_run_id=1842832884#step:8:7

```
Checking dist/mlflow_skinny-1.13.2.dev0-py3-none-any.whl: FAILED
  `long_description` has syntax errors in markup and would not be rendered on PyPI.
    line 16: Warning: Bullet list ends without a blank line; unexpected unindent.
```

Looks like we need to insert a blank line here.

Comment (harupy) [reply README_SKINNY.rst:15]: or we could concatenate `README_SKINNY.rs` and `README.rst` with a new line.

Comment (dbczumar) [reply README_SKINNY.rst:15]: Let’s add the blank line. Sorry for the inconvenience here

Comment (harupy) [reply README_SKINNY.rst:15]: DONE!

**Review by harupy — APPROVED** [review]

**Pull request merged**
